### ISSUE TITLE
Build trouble in VSCode

### ISSUE THREAD

**Issue by vsoch** [flux-framework/PerfFlowAspect#112]
Hi! I'm trying to follow the logic in the GitHub workflow and I have this Dockerfile:

```dockerfile
FROM ubuntu:20.04

LABEL maintainer="Vanessasaurus <@vsoch>"

# Match the default user id for a single system so we aren't root
ARG USERNAME=vscode
ARG USER_UID=1000
ARG USER_GID=1000
ENV USERNAME=${USERNAME}
ENV USER_UID=${USER_UID}
ENV USER_GID=${USER_GID}
ENV DEBIAN_FRONTEND=noninteractive

RUN apt-get update && \
    apt install -y clang llvm-dev libjansson-dev libssl-dev \
    mpi-default-dev wget bison flex make cmake mpich python3 python3-pip sudo llvm-12 
    # Note I didn't get it working with just mpich!
#    openmpi-bin openmpi-common libopenmpi-dev libgtk2.0-dev

# install cuda 12.1 (note uncomment this if you can support on your development machine!)
RUN wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/cuda-ubuntu2004.pin && \
    mv cuda-ubuntu2004.pin /etc/apt/preferences.d/cuda-repository-pin-600 && \
    wget https://developer.download.nvidia.com/compute/cuda/12.1.0/local_installers/cuda-repo-ubuntu2004-12-1-local_12.1.0-530.30.02-1_amd64.deb && \
    dpkg -i cuda-repo-ubuntu2004-12-1-local_12.1.0-530.30.02-1_amd64.deb && \
    cp /var/cuda-repo-ubuntu2004-12-1-local/cuda-*-keyring.gpg /usr/share/keyrings/ && \
    apt-get update && \
    apt-get -y install cuda && \
    ln -s -T /usr/bin/make /usr/bin/gmake

ENV PATH=/usr/local/cuda-12.1/bin:$PATH

# Python helpers
RUN python3 -m pip install --upgrade pip pytest setuptools flake8 rstfmt black
#    sudo update-alternatives --install /usr/bin/llvm-config llvm-config /usr/bin/llvm-config-12 200

RUN ldconfig

# Add the group and user that match our ids
RUN groupadd -g ${USER_GID} ${USERNAME} && \
    adduser --disabled-password --uid ${USER_UID} --gid ${USER_GID} --gecos "" ${USERNAME} && \
    echo "${USERNAME} ALL=(ALL) NOPASSWD: ALL" > /etc/sudoers
USER $USERNAME
```

And then to build:

```console
export PYTHONPATH=$PWD/src/python:$PYTHONPATH
cd src/c
mkdir -p build install
cd build
export CMAKE_OPTS="-DCMAKE_CXX_COMPILER=clang++ -DLLVM_DIR=/usr/lib/llvm-12/cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=../install"
cmake ${CMAKE_OPTS} ..
```

But the build fails - here is the issue

```console
Determining if the C compiler works failed with the following output:
Change Dir: /workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/gmake cmTC_599b9/fast && No such file or directory
Generator: execution of make failed. Make command was: /usr/bin/gmake cmTC_599b9/fast && 


The MPI test test_mpi for CXX in mode normal failed to compile with the following output:
Change Dir: /workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/gmake cmTC_21bda/fast && /usr/bin/gmake -f CMakeFiles/cmTC_21bda.dir/build.make CMakeFiles/cmTC_21bda.dir/build
gmake[1]: Entering directory '/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_21bda.dir/test_mpi.cpp.o
/usr/bin/clang++   -isystem /usr/include/x86_64-linux-gnu/mpich  -fPIE   -flto=auto -ffat-lto-objects -o CMakeFiles/cmTC_21bda.dir/test_mpi.cpp.o -c /workspaces/PerfFlowAspect/src/c/build/CMakeFiles/FindMPI/test_mpi.cpp
clang: error: unsupported argument 'auto' to option 'flto='
clang: warning: optimization flag '-ffat-lto-objects' is not supported [-Wignored-optimization-argument]
gmake[1]: *** [CMakeFiles/cmTC_21bda.dir/build.make:66: CMakeFiles/cmTC_21bda.dir/test_mpi.cpp.o] Error 1
gmake[1]: Leaving directory '/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeTmp'
gmake: *** [Makefile:121: cmTC_21bda/fast] Error 2
```

Is there a flag or something I;m missing on a path or a version error? I'm using ubuntu 20.04 and mpich like the workflow file. Thanks!

**Comment by tpatki** [comment]
@vsoch We've mostly tested this on lassen, and we use the `clang-10.0.1-gcc-8.3.1` module there, which I believe is clang built with gcc bindings, I wonder if that's what we need to install here. 

I didn't see (or missed your actual `make` command), hopefully you're not using a parallel make, because that won't work with the C dependencies we have.  Can you post the verbose output from `make` here? 

I am assuming your code is in `cmTC_599b9/fast` ? Not sure what the directory is pointing to. It seems like it is missing some directory as per this, not sure what its looking for: 
```
Run Build Command(s):/usr/bin/gmake cmTC_599b9/fast && No such file or directory
``` 

Also see some incorrect clang flags, not sure how those got added, they're not in our codebase:

```
clang: error: unsupported argument 'auto' to option 'flto='
clang: warning: optimization flag '-ffat-lto-objects' is not supported 
```

Happy to help debug over a call sometime.

**Comment by slabasan** [comment]
@vsoch Can you try with llvm-10 instead of llvm-12? I think this is the last piece that is on a different version. It was tricky to find a combination of llvm, clang, and cuda versions to work for our CI.

**Comment by vsoch** [comment]
This is really strange - when I build the exact same container from VSCode, I get a different error:

```
-- Setting build type to "Debug"
-- CMAKE_INSTALL_RPATH = /workspaces/PerfFlowAspect/src/c/install/lib
-- Could NOT find MPI_CXX (missing: MPI_CXX_WORKS) 
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find MPI (missing: MPI_CXX_FOUND) (found version "3.1")
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindMPI.cmake:1688 (find_package_handle_standard_args)
  test/CMakeLists.txt:10 (find_package)


-- Configuring incomplete, errors occurred!
See also "/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeOutput.log".
See also "/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeError.log".
```

I tried as root (as I did above) and the vscode user (same issue!) I think this was an original error I saw, so maybe we should debug from here. So it's telling me MPI isn't installed - but I definitely installed mpich. Here are the paths to what I found in /usr with mpi:

```
$ find . -name *mpi*.so
./lib/x86_64-linux-gnu/libnvidia-ptxjitcompiler.so
./lib/x86_64-linux-gnu/libmpich.so
./lib/x86_64-linux-gnu/libompitrace.so
./lib/x86_64-linux-gnu/libmpichfort.so
./lib/x86_64-linux-gnu/libmpichcxx.so
./lib/x86_64-linux-gnu/openmpi/lib/libompitrace.so
./lib/x86_64-linux-gnu/openmpi/lib/openmpi3/libompi_dbg_msgq.so
./lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_schizo_ompi.so
./lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_io_ompio.so
./lib/x86_64-linux-gnu/openmpi/lib/libmca_common_ompio.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi_java.so
./lib/x86_64-linux-gnu/openmpi/lib/ompi_monitoring_prof.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi_usempi_ignore_tkr.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi_mpifh.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi.so
./lib/x86_64-linux-gnu/openmpi/lib/libmpi_usempif08.so
./lib/x86_64-linux-gnu/libmca_common_ompio.so
./lib/x86_64-linux-gnu/libmpi_java.so
./lib/x86_64-linux-gnu/libmpi_usempi_ignore_tkr.so
./lib/x86_64-linux-gnu/open-coarrays/openmpi/lib/libcaf_mpi.so
./lib/x86_64-linux-gnu/open-coarrays/openmpi/lib/libcaf_openmpi.so
./lib/x86_64-linux-gnu/libmpi_mpifh.so
./lib/x86_64-linux-gnu/libmpi_cxx.so
./lib/x86_64-linux-gnu/libmpi.so
./lib/x86_64-linux-gnu/libcaf_openmpi.so
./lib/x86_64-linux-gnu/libmpi_usempif08.so
./lib/x86_64-linux-gnu/libmpi++.so
```
What should I try?

**Comment by slabasan** [comment]
Let's try `./lib/x86_64-linux-gnu/libmpich.so` and `./lib/x86_64-linux-gnu/libmpichcxx.so`

**Comment by vsoch** [comment]
How should I do that with cmake (sorry I don't use it a lot!) I was messing around with something like:

```
cmake -DCMAKE_CXX_COMPILER=clang++ -DMPI_DIR=/lib/x86_64-linux-gnu/ -DLLVM_DIR=/usr/lib/llvm-10/cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_INSTALL_PREFIX=../install ..
```

Also note that I was installing cuda as you do in your GitHub workflow - should I not do that (maybe it's mucking something up?) I think it's weird I see openmpi and mpich, like maybe the nvidia installed grabbed the first?

**Comment by slabasan** [comment]
```
cmake \
-DCMAKE_CXX_COMPILER=clang++ \
-DMPI_CXX=./lib/x86_64-linux-gnu/libmpichcxx.so \
-DMPI_C=./lib/x86_64-linux-gnu/libmpich.so \
-DLLVM_DIR=/usr/lib/llvm-10/cmake \
-DCMAKE_BUILD_TYPE=Debug \
-DCMAKE_INSTALL_PREFIX=../install \
..
```

And it was a big headache when we brought in cuda into the CI environment. I don't think it should affect your build, especially if we're specific about which one we want the build system to use. IIRC the errors I was running into was making sure it picked up cuda-12, and not the default cuda. Needed to remove old drivers first instead of just replacing them.

**Comment by vsoch** [comment]
ah the above makes sense! So I tried it for each of llvm-10 and llvm-12 and the error is the same (I'm also not sure if the relative path dot in front of lib should be removed so I tried that too). It's giving the same error. Could this be that it's finding it but it's the wrong version? Likely on the cluster there is specific version of mpich and here we are just getting what ubuntu provides. E.g., I'm looking at ` (found version "3.1")` so does that mean it found something?

**Comment by slabasan** [comment]
In the CMake output, is it finding the paths you specified? 
```
-- Found MPI_C: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so (found version "3.1") 
-- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
```

And when you run `VERBOSE=1 make`, are you seeing that the build has picked up other mpich libraries?

**Comment by slabasan** [comment]
Hi @vsoch, let me take a look at my CMakeCache.txt and confirm the variable to set for the MPI libraries, looks like MPI_C isn't the right one.

**Comment by vsoch** [comment]
I don't believe so - I assumed it was pulled in by the cuda install. Shall I try the build without those early cuda steps and check if it's there?

**Comment by vsoch** [comment]
okay this gives us some insights! It seems it _was_ targeting openmpi, because now it's angry that it's missing! :laughing: 

```
CMake Error in /workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeTmp/CMakeLists.txt:
  Imported target "MPI::MPI_C" includes non-existent path

    "/usr/lib/x86_64-linux-gnu/openmpi/include/openmpi"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error at /usr/share/cmake-3.16/Modules/FindMPI.cmake:1194 (try_compile):
  Failed to generate test project build system.
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindMPI.cmake:1245 (_MPI_try_staged_settings)
  /usr/share/cmake-3.16/Modules/FindMPI.cmake:1505 (_MPI_check_lang_works)
  test/CMakeLists.txt:10 (find_package)


-- Configuring incomplete, errors occurred!
See also "/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeOutput.log".
See also "/workspaces/PerfFlowAspect/src/c/build/CMakeFiles/CMakeError.log".
```

**Comment by slabasan** [comment]
Hi @vsoch, ok I was able to successfully build PerfFlowAspect in a docker container. Our steps are similar, so not sure why your build is still looking for openmpi. 

This was my docker setup (I ran my commands manually in a new ubuntu 20.04 instance:
```
apt-get update
apt-get install -y clang llvm-dev libjansson-dev libssl-dev wget bison flex make cmake python3 python3-pip llvm-12 
apt-get install -y mpich

wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/cuda-ubuntu2004.pin && \
mv cuda-ubuntu2004.pin /etc/apt/preferences.d/cuda-repository-pin-600 && \
wget https://developer.download.nvidia.com/compute/cuda/12.1.0/local_installers/cuda-repo-ubuntu2004-12-1-local_12.1.0-530.30.02-1_amd64.deb && \
dpkg -i cuda-repo-ubuntu2004-12-1-local_12.1.0-530.30.02-1_amd64.deb && \cp /var/cuda-repo-ubuntu2004-12-1-local/cuda-*-keyring.gpg /usr/share/keyrings/ && \
apt-get update && \
apt-get -y install cuda && \
ln -s -T /usr/bin/make /usr/bin/gmake

export PATH=/usr/local/cuda-12.1/bin:$PATH
```

I can confirm I do not have an openmpi in `/usr/lib/x86_64-linux-gnu/openmpi/`. I'll leave my docker instance up, let me know if I can look up anything else, so we can compare.

This build line works for me: `$ cmake -DCMAKE_CXX_COMPILER=clang++  -DLLVM_DIR=/usr/lib/llvm-10/cmake ..`
```
-- The C compiler identification is GNU 9.4.0
-- The CXX compiler identification is Clang 10.0.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/clang++
-- Check for working CXX compiler: /usr/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Using default build type: "Debug"
-- CMAKE_INSTALL_RPATH = /usr/local/lib
-- Found BISON: /usr/bin/bison (found version "3.5.1") 
-- Found FLEX: /usr/bin/flex (found version "2.6.4") 
-- Found OpenSSL: /usr/lib/x86_64-linux-gnu/libcrypto.so (found version "1.1.1f")  
-- Found MPI_C: /usr/lib/x86_64-linux-gnu/libmpich.so (found version "3.1") 
-- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/libmpichcxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found CUDA: /usr/local/cuda-12.1 (found version "12.1") 
-- Adding CXX unit tests
--  [*] Adding test: smoketest
--  [*] Adding test: smoketest2
--  [*] Adding test: smoketest3
--  [*] Adding test: smoketest_class
--  [*] Adding test: smoketest_MPI
--  [*] Adding test: smoketest_MT
--  [*] Adding test: smoketest_cuda
-- Config Dir: 
-- PerfFlowAspect version: "0.1.0"
-- Configuring done
-- Generating done
-- Build files have been written to: /root/PerfFlowAspect/src/c/build
```

**Comment by vsoch** [comment]
Trying now! I'm worried this is related to the VSCode environment, because I had a variant working in just docker before. Will report back!

**Comment by vsoch** [comment]
okay - failed in the .devcontainer. @slabasan do you use/have VSCode and could you reproduce the devcontainer setup? Basically you can put the Dockerfile alongside this `devcontainer.json`

```
{
    "name": "PerfFlowAspect Development Environment",
    "dockerFile": "Dockerfile",
    "context": "../",

    "customizations": {
      "vscode": {
        "settings": {
          "terminal.integrated.defaultProfile.linux": "bash"
        }
      }
  },
  "postStartCommand": "git config --global --add safe.directory /workspaces/PerfFlowAspect"
}
```
and open in VSCode, it will ask you if you want to reopen in the container, and if not,  you can do View -> Command Palette -> Rebuild in Container. The extension is called "Dev Containers" by Microsoft.

I'm staring to think the build is fine and the specific developer environment is somehow wonky, but I've never seen this before! If VSCode doesn't work, I'll provide a standard Dockerfile instead for a developer environment - not as smooth but better than nothing!

**Comment by slabasan** [comment]
I’ve never used VSCode, but happy to try and reproduce. Do you have beginning steps I can follow?

**Comment by vsoch** [comment]
Yeah! It's fairly simple - you basically need VSCode, docker, and the extension. We have a small guide in the flux repository: https://github.com/flux-framework/flux-core/blob/master/vscode.md and let me know if you have specific questions (glad to help, and grateful for you trying to reproduce)!

**Comment by vsoch** [comment]
okay - I've updated the title to reflect the issues are in VSCode - the regular build with a vanilla Dockerfile works great. To not slow down progress, I'm going to leave this issue open but continue with a development environment that just uses a container:

https://github.com/flux-framework/PerfFlowAspect/pull/114

This was how I did things before we had VSCode - you basically can just build in the container, or bind your code to the container (and make changes on the host that persist in the container). If we ever figure this out or it magically works, great! But if not, we should not be held up because of it.
